src/nvim/os/env.c

```diff
@@ -151,25 +151,39 @@ int os_unsetenv(const char *name)
 char *os_getenvname_at_index(size_t index)
 {
 #ifdef _WIN32
-  // Check if index is inside the environ array and is not the last element.
-  for (size_t i = 0; i <= index; i++) {
-    if (_wenviron[i] == NULL) {
-      return NULL;
-    }
-  }
-  wchar_t *utf16_str = _wenviron[index];
-  char *utf8_str;
-  int conversion_result = utf16_to_utf8(utf16_str, &utf8_str);
-  if (conversion_result != 0) {
-    EMSG2("utf16_to_utf8 failed: %d", conversion_result);
+  wchar_t *env = GetEnvironmentStringsW();
+  if (!env) {
     return NULL;
   }
-  size_t namesize = 0;
-  while (utf8_str[namesize] != '=' && utf8_str[namesize] != NUL) {
-    namesize++;
+
+  // GetEnvironmentStrings return environment block of the following format:
+  //
+  // var1=value1\0var2=value2\0...varN=valueN\0\0
+  //
+  char *name = NULL;
+  size_t current_index = 0;
+  for (wchar_t *it = env; *it != L'\0' || *(it + 1) != L'\0'; it++) {
+    if (index == current_index) {
+      char *utf8_str;
+      int conversion_result = utf16_to_utf8(it, &utf8_str);
```

Comment:
@erw7 the old code did one call to `utf16_to_utf8()` on the whole `env` string. Is your change a performance optimization?